src/metabase/query_processor/middleware/annotate.clj

```diff
@@ -519,11 +514,15 @@
 
 (defn- flow-field-metadata
   "Merge information about fields from `source-metadata` into the returned `cols`."
-  [source-metadata cols]
-  (let [ref->metadata (u/key-by (comp u/field-ref->key :field_ref) source-metadata)]
+  [source-metadata cols dataset?]
+  (let [index           (fn [col] (or (:id col) (:name col "")))
```

Comment:
This is the actual fix for this issue: using `id` then `name` to match the outer select field to the inner select field. Because if there are aggregations at both levels, the field_refs are _NOT_ unique (both being `[:aggregation 0]` etc) and those are matched when they should not be.